flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/StringToTimestampCastRule.java

```diff
@@ -48,12 +52,33 @@ public String generateExpression(
             String inputTerm,
             LogicalType inputLogicalType,
             LogicalType targetLogicalType) {
+        int targetPrecision = LogicalTypeChecks.getPrecision(targetLogicalType);
+        if (inputLogicalType.is(LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE)
```

Comment:
reminder: Whoever merges second, we should extract this to a method: https://github.com/apache/flink/pull/18582/files#diff-50ec7a5a77278fff0d10a781f5f41f84f581c7bf3617a722b003de7ca01a45e8R61

Reply:
Actually, this was wrong, because i'm checking input type as well. I modified it